Sources/TuistDependencies/Mappers/ExternalProjectsPlatformTreeShakerGraphMapper.swift

```diff
@@ -0,0 +1,128 @@
+import Foundation
+import TuistCore
+import TuistGraph
+
+/**
+ When Swift Packages don't declare the platforms that they support, the Swift Package Manager defaults the value
+ to 'support all the platforms'. This default behaviour is inherited into the Xcode projects that we generate off the packages
+ and that causes compilation issues. Xcode must resolve this issue at build-time by cascading the platform requirements
+ down from nodes in the graph that are closer to the root. This is a behaviour that we need to copy over to Tuist. In our case
+ the logic is executed at generation time.
+ */
+public final class ExternalProjectsPlatformTreeShakerGraphMapper: GraphMapping {
+    public init() {}
+
+    public func map(graph: Graph) async throws -> (Graph, [TuistCore.SideEffectDescriptor]) {
+        // If the project has no external dependencies we skip this.
+        if graph.projects.values.first(where: { $0.isExternal }) == nil {
+            return (graph, [])
+        }
+
+        var graph = graph
+
+        /**
+         One traverse gives us a dictionary that maps graph targets to the platforms they should support
+         */
+        let targetsToPlatformsDictionary = targetsToPlatformsDictionary(graph: graph)
+
+        graph.targets = Dictionary(uniqueKeysWithValues: graph.targets.map { projectPath, projectTargets in
+            let project = graph.projects[projectPath]!
+            let projectTargets = Dictionary(uniqueKeysWithValues: projectTargets.map { targetName, target in
+                let graphTarget = GraphTarget(path: projectPath, target: target, project: project)
+                var target = target
+
+                /**
+                 We only include the destinations whose platform is included in the list of the target supported platforms.
+                 */
+                if let targetFilteredPlatforms = targetsToPlatformsDictionary[graphTarget] {
+                    target.destinations = target.destinations.filter { destination in
+                        targetFilteredPlatforms.contains(destination.platform)
+                    }
+                }
+                return (targetName, target)
+            })
+            return (projectPath, projectTargets)
+        })
+        return (graph, [])
+    }
+
+    private func targetsToPlatformsDictionary(graph: Graph) -> [GraphTarget: Set<Platform>] {
```

Comment:
@waltflanagan / @kwridan do you think it's necessary to traverse the entire graph, or would it be sufficient if we obtain the platforms from the first nodes of the project that depend on a external dependency?

Reply:
assuming that the targets are well formed (define just the platforms they actually support) i think it will be sufficient to only compute this for targets with direct external dependencies.  those targets should be enough to gather the concrete platform requirements for all external dependencies.  I think any other complexities from filters would be from the leaf targets through the graph of external dependencies and their conditions.

Reply:
I'll adjust the code accordingly. We can iterate on it as needed.